packages/store/addon/-private/ts-interfaces/minimum-serializer-interface.ts

```diff
@@ -0,0 +1,280 @@
+   * when normalizing included records. If mixing serializer usage in this way
+   * we recommend implementing this method, but caution that it may lead
+   * to unexpected mixing of formats.
+   *
+   * This method may also be called when normalizing embedded relationships when
+   * using the `EmbeddedRecordsMixin`. If using this mixin in a serializer in
+   * your application we recommend implementing this method, but caution that
+   * it may lead to unexpected mixing of formats.
+   *
+   * @method normalize [OPTIONAL]
+   * @public
+   * @optional
+   * @param {ShimModelClass} schema - An object with methods for accessing information about
+   *  the type, attributes and relationships of the primary type associated with the request.
+   * @param {JSONObject} rawPayload - Some raw JSON data to be normalized into a `JSON:API` resource.
+   * @param {string} [prop] - When called by the `EmbeddedRecordsMixin` this param will be the
+   *  property at which the object provided as rawPayload was found.
+   * @returns {SingleResourceDocument} - A `JSON:API` document containing a single `resource` as
+   * its primary data.
+   */
+  normalize?(schema: ShimModelClass, rawPayload: JSONObject, prop?: string): SingleResourceDocument;
+
+  /**
+   * When using `JSONAPIAdapter` or `RESTAdapter` this method is called
+   * by `adapter.updateRecord` and `adapter.createRecord` if `Serializer.serializeIntoHash`
+   * is not implemented.
+   *
+   * You can use this method to customize the root keys serialized into the payload.
+   * The hash property should be modified by reference.
+   *
+   * For instance, your API may expect resources to be keyed by underscored type in the payload:
+   *
+   * ```js
+   * {
+   *   _user: {
+   *     type: 'user',
+   *     id: '1'
+   *   }
+   * }
+   * ```
+   *
+   * Which when using these adapters can be achieved by implementing this method similar
+   * to the following:
+   *
+   * ```js
+   * serializeIntoHash(hash, ModelClass, snapshot, options) {
+   *   hash[`_${snapshot.modelName}`] = this.serialize(snapshot, options).data;
+   * }
+   * ```
+   *
+   * @method serializeIntoHash [OPTIONAL]
+   * @public
+   * @optional
+   * @param hash - an top most object of the request payload onto
+   *  which to append the serialized record
+   * @param {ShimModelClass} schema - An object with methods for accessing information about
+   *  the type, attributes and relationships of the primary type associated with the request.
+   * @param {Snapshot} snapshot - A Snapshot for the record to serialize
+   * @param options
+   * @returns {void}
+   */
+  serializeIntoHash?(hash: object, schema: ShimModelClass, snapshot: Snapshot, options?: OptionsHash): void;
+
+  /**
+   * This method is called by `store.pushPayload` and should be implemented if
+   * you want to use that method.
+   *
+   * It is recommended to use `store.push` over `store.pushPayload` after normalizing
```

Comment:
really? Why?

Reply:
There's a few layers to the motivations for this:

Brittleness

- users don't have clarity around which serializer's `pushPayload` is going to be called and the serializer is not given the context of what the "primary type" for a given pushPayload call is expected to be. This makes the normalization "blind" to what format it is normalizing from.
- the convenience REST serializer expects a format that differs from what would be expected normalizing any other payload.

Features

- users often want access to the records they have pushed, which they get from `push` but not `pushPayload`

Long Term Support

- We've eyeballed both this serializer method and the associated store method as a candidate for deprecation for a long time given it (A) is redundant (B) is not part of any primary APIs (C) encourages a non-linear flow for processing data and (D) is one of several points that force the store to be aware of the specifics of serialization and normalization which is not it's concern

Reply:
I would be prepared for people to take this statement as "`pushPayload` is bad, and you shouldn't use it". AFAIK they're both useful for different things (single record where you get a reference back, vs. complex payload where there is no "primary" thing to return as a record).

This smells a little bit like an attempt to pre-deprecate, which our community has suffered from in the past (e.g., "controllers are dead").

Reply:
@mike-north it is what it smells like :( We intend to deprecate this but haven't had time to work on the RFC https://github.com/emberjs/rfcs/issues/357

Now that more contributors are getting involved though maybe there are hands for this. @snewcomer @code0100fun would either of you be interested in taking on this deprecation RFC so that we don't leave this guidance hanging in a weird state?

Reply:
In the short term, I'd leave this out of the docs until such time that we're ready to do this the right way.

Deprecation is the tool we use to steer people away from APIs we intend to remove, and we should use it for things like this

Reply:
I updated the recommendation a bit here to be less "soft-deprecation" and more "why/where to steer away"